backend/services/user_groups.go

```diff
@@ -0,0 +1,421 @@
+// Copyright 2022, Yahoo Inc.
+// Licensed under the terms of the MIT. See LICENSE file in project root for terms.
+
+package services
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"math"
+	"strings"
+	"time"
+	"unicode"
+
+	"github.com/theparanoids/ashirt-server/backend"
+	"github.com/theparanoids/ashirt-server/backend/database"
+	"github.com/theparanoids/ashirt-server/backend/dtos"
+	"github.com/theparanoids/ashirt-server/backend/models"
+	"github.com/theparanoids/ashirt-server/backend/policy"
+	"github.com/theparanoids/ashirt-server/backend/server/middleware"
+
+	sq "github.com/Masterminds/squirrel"
+)
+
+type CreateUserGroupInput struct {
+	Name      string
+	Slug      string
+	UserSlugs []string
+}
+
+type ModifyUserGroupInput struct {
+	Name          string
+	Slug          string
+	UsersToAdd    []string
+	UsersToRemove []string
+}
+
+type ListUserGroupsForAdminInput struct {
+	UserGroupFilter
+	Pagination
+	IncludeDeleted bool
+}
+
+type ListUserGroupsForOperationInput struct {
+	Pagination
+	UserGroupFilter
+	OperationSlug string
+}
+
+type userGroupAndRole struct {
+	models.UserGroup
+	Role policy.OperationRole `db:"role"`
+}
+
+type ListUserGroupsInput struct {
+	Query          string
+	IncludeDeleted bool
+	OperationSlug  string
+}
+
+func (i ModifyUserGroupInput) validateUserGroupInput() error {
+	if i.Slug == "" {
+		return backend.MissingValueErr("Slug")
+	}
+	return nil
+}
+
+func (i CreateUserGroupInput) validateUserGroupInput() error {
+	if i.Slug == "" {
+		return backend.MissingValueErr("Slug")
+	}
+	if i.Name == "" {
+		return backend.MissingValueErr("Name")
+	}
+	return nil
+}
+
+func AddUsersToGroup(tx *database.Transactable, userSlugs []string, groupID int64) error {
+	if len(userSlugs) > 0 {
+		questionMarks := "("
+
+		interfaceSlice := make([]interface{}, len(userSlugs)+1)
+		for i, v := range userSlugs {
+			questionMarks += "?, "
+			interfaceSlice[i] = v
+		}
+
+		questionMarks = strings.TrimSuffix(questionMarks, ", ")
+		questionMarks += ")"
+
+		sqlStatement := fmt.Sprintf(`INSERT IGNORE INTO group_user_map(user_id, group_id)
+					SELECT users.id, user_groups.id
+					FROM users, user_groups
+					WHERE users.slug in %s and user_groups.id = ?;`, questionMarks)
+
+		interfaceSlice[len(userSlugs)] = groupID
+		err := tx.Exec(sq.Expr(sqlStatement, interfaceSlice...))
+
+		if err != nil {
+			return backend.WrapError("Unable to add users to group", backend.DatabaseErr(err))
+		}
+		return nil
+	}
+
+	return nil
+}
```

Comment:
```suggestion
func AddUsersToGroup(tx *database.Transactable, userSlugs []string, groupID int64) error {
	if len(userSlugs) == 0 {
		return nil
	}

	peopleToAdd := sq.Select("id", strconv.FormatInt(groupID, 10)).
		From("users").
		Where(sq.Eq{"slug": userSlugs})

	insertQuery := sq.
		Insert("group_user_map").
		Options("Ignore").
		Columns("user_id", "group_id").Select(peopleToAdd)

	err := tx.Exec(insertQuery)

	if err != nil {
		return backend.WrapError("Unable to add users to group", backend.DatabaseErr(err))
	}
	return nil
}
```

Here's what's going on:

We can't use the  normal Insert or BatchInsert functions because those are more designed for normal insert cases, which this isn't. Instead, we're going to use what those use: the Squirrel InsertBuilder and tx.Exec.

Let's start on the insertQuery line: We're going to start a normal insert, but we're going to tack on the ignore option, which should get us `INSERT IGNORE INTO`.We'll then specify the columns as usual, but instead of specifying the values, we'll defer to a select statement.
Back on the peopleToAdd line, we're going to select the user's id for a given slug, but we're _also_ going to select a fixed value so that we end up with a series of (user_id/user_group_id) pairs.

Adding those two things together, we get a single sql query (unexecuted) that should get us what we want. tx.Exec makes it official, and everything else is basically the same.

Really, this is just doing what you hand coded. The only "significant" change is that I'm selecting a constant value, rather than searching the user_group table for an id I already know. I'd imagine that behind the scenes sql will somewhat optimize that query, so it wouldn't  have been a slow solution, just a strange one.

Reply:
oh nice - yeah this is much cleaner